src/VisualStudio/Core/Def/Implementation/FindReferences/AsyncFindReferencesPresenter.DataSource.cs

```diff
@@ -0,0 +1,180 @@
+            {
+                var subscription = new Subscription(_cancellationTokenSource, sink);
+                _subscriptions.Add(subscription);
+
+                sink.AddFactory(this, removeAllFactories: true);
+
+                return subscription;
+            }
+
+            #endregion
+
+            #region EditorFindReferencesContext overrides.
+
+            public override void OnStarted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = false;
+                }
+            }
+
+            public override void OnCompleted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = true;
+                }
+            }
+
+            public async override void OnReferenceFound(INavigableItem definition, INavigableItem reference)
+            {
+                var workspace = reference.Document.Project.Solution.Workspace as VisualStudioWorkspaceImpl;
+                if (workspace == null)
+                {
+                    return;
+                }
+
+                var projectGuid = workspace.GetHostProject(reference.Document.Project.Id)?.Guid;
+                if (projectGuid == null)
+                {
+                    return;
+                }
+                
+                try
+                {
+                    // We're told about this reference synchronously, but we need to get the 
+                    // SourceText for the reference's Document so that we can determine things
+                    // like it's line/column/text.  We don't want to block this method getting
+                    // that data, so instead we just fire off the async work to get the text
+                    // and use it.  Because we're starting some async work, let the test harness
+                    // know so that it doesn't verify results until this completes.
+                    using (var token = _asyncListener.BeginAsyncOperation(nameof(OnReferenceFound)))
+                    {
+                        await OnReferenceFoundAsync(definition, reference, projectGuid.Value).ConfigureAwait(false);
+                    }
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                }
+            }
+
+            private async Task OnReferenceFoundAsync(
+                INavigableItem definition, INavigableItem reference, Guid projectGuid)
+            {
+                var sourceText = await reference.Document.GetTextAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
+
+                lock (_gate)
+                {
+                    // Once we can make the new entry, add it to our list.
+                    _entries = _entries.Add(new TableEntry(definition, reference, projectGuid, sourceText));
```

Comment:
Is this `O(n^2)`? Should this be `ImmutableList` or something else that you're calling .Add on to allow for cheaper adds?